File: task_3_3_shared_variable_mutex.py
"""
Task 3.3 - Approach 3: Shared Variable with Synchronization
This version uses multiprocessing and shared state with a lock.

Design:
- Use a shared counter across processes
- Each process computes a local result first
- Then each process updates the shared counter inside a locked section

Why this changes Task 3:
- It preserves the shared-variable-with-synchronization model
- It also allows real speedup for a CPU-bound workload
- The lock is held only briefly during final update

Measurement:
- Same setup as other approaches
- Compare execution time and speedup
- Show how multiprocessing can improve performance over single-process baseline
"""

import random
import time
import math
from multiprocessing import Process, Manager


def worker(process_id, samples_per_process, shared_counter, lock):
    """
    Worker function for each process.
    Each process computes its local count first, then updates shared counter.
    """
    random.seed(42 + process_id)
    local_inside_circle = 0
    
    for _ in range(samples_per_process):
        x = random.uniform(-1, 1)
        y = random.uniform(-1, 1)
        if x**2 + y**2 <= 1:
            local_inside_circle += 1
    
    # Update shared counter under lock once per process
    with lock:
        shared_counter.value += local_inside_circle


def monte_carlo_pi_multi_process_mutex(num_samples, num_processes):
    """
    Multi-process Monte Carlo estimation of Pi using a shared counter and lock.
    """
    samples_per_process = num_samples // num_processes
    manager = Manager()
    shared_counter = manager.Value('i', 0)
    lock = manager.Lock()
    processes = []
    
    start_time = time.time()
    
    for process_id in range(num_processes):
        process = Process(
            target=worker,
            args=(process_id, samples_per_process, shared_counter, lock)
        )
        processes.append(process)
        process.start()
    
    for process in processes:
        process.join()
    
    elapsed_time = time.time() - start_time
    estimated_pi = (shared_counter.value / num_samples) * 4
    return estimated_pi, elapsed_time


def measure_performance(num_samples, process_counts, num_runs=3):
    """
    Measure the performance of the shared-variable multiprocessing approach.
    """
    baseline_time = None
    results = {}
    
    print(f"Running multiprocessing approach (shared counter with lock) with {num_samples:,} samples...")
    print(f"Running {num_runs} times per configuration for averaging...\n")
    
    for num_processes in process_counts:
        times = []
        pi_estimates = []
        
        print(f"Testing with {num_processes:2d} process(es)...", end=" ", flush=True)
        
        for run in range(num_runs):
            estimated_pi, elapsed_time = monte_carlo_pi_multi_process_mutex(num_samples, num_processes)
            times.append(elapsed_time)
            pi_estimates.append(estimated_pi)
        
        avg_time = sum(times) / len(times)
        min_time = min(times)
        max_time = max(times)
        avg_pi = sum(pi_estimates) / len(pi_estimates)
        accuracy = (1 - abs(avg_pi - math.pi) / math.pi) * 100
        
        if num_processes == 1:
            baseline_time = avg_time
        
        speedup = baseline_time / avg_time if baseline_time else 1.0
        efficiency = (speedup / num_processes) * 100
        
        results[num_processes] = {
            'avg_time': avg_time,
            'min_time': min_time,
            'max_time': max_time,
            'all_times': times,
            'estimated_pi': avg_pi,
            'accuracy': accuracy,
            'speedup': speedup,
            'efficiency': efficiency
        }
        
        print(f"Avg time: {avg_time:.6f}s, Speedup: {speedup:.4f}x, Efficiency: {efficiency:.2f}%")
    
    return results, baseline_time


if __name__ == "__main__":
    num_samples = 10**7
    process_counts = [1, 2, 4, 8]
    
    results, baseline_time = measure_performance(num_samples, process_counts, num_runs=3)
    
    print("\n" + "="*90)
    print("MULTI-PROCESS (SHARED VARIABLE WITH LOCK) RESULTS")
    print("="*90)
    print(f"{'Processes':<12} {'Avg Time (s)':<15} {'Speedup':<12} {'Efficiency':<12} {'Accuracy':<12}")
    print("-"*90)
    
    for num_processes in process_counts:
        result = results[num_processes]
        print(f"{num_processes:<12} {result['avg_time']:<15.6f} {result['speedup']:<12.4f}x "
              f"{result['efficiency']:<12.2f}% {result['accuracy']:<12.4f}%")
    
    print("="*90)
    print(f"\nTrue Pi: {math.pi:.8f}")
    print(f"Estimated Pi (final run): {results[process_counts[-1]]['estimated_pi']:.8f}")
    print(f"\nBaseline time (1 process): {baseline_time:.6f} seconds")
    print(f"Maximum speedup achieved: {results[max(results.keys())]['speedup']:.4f}x")
    print("\nThis version uses a shared counter plus a lock across processes, but only updates the shared state once per process.")
    print("It therefore still shows speedup while preserving the shared-variable-with-synchronization pattern.")

File: task_3_3_shared_variable_mutex.c
#include <stdio.h>
#include <stdlib.h>
#include <pthread.h>
#include <time.h>
#include <math.h>

typedef struct {
    long long samples;
    unsigned int seed;
    long long *shared_inside;
    pthread_mutex_t *mutex;
} ThreadData;

static double get_seconds(void) {
    struct timespec ts;
    clock_gettime(CLOCK_MONOTONIC, &ts);
    return ts.tv_sec + ts.tv_nsec * 1e-9;
}

void *worker(void *arg) {
    ThreadData *data = (ThreadData *)arg;
    unsigned int state = data->seed;
    for (long long i = 0; i < data->samples; ++i) {
        double x = (double)rand_r(&state) / RAND_MAX * 2.0 - 1.0;
        double y = (double)rand_r(&state) / RAND_MAX * 2.0 - 1.0;
        if (x*x + y*y <= 1.0) {
            // Each thread updates the shared variable directly with mutex protection
            pthread_mutex_lock(data->mutex);
            *data->shared_inside += 1;
            pthread_mutex_unlock(data->mutex);
        }
    }
    return NULL;
}

int main(void) {
    const long long num_samples = 10000000LL;
    const int num_runs = 3;
    const int thread_counts[] = {1, 2, 4, 8, 16, 32};
    const int count_len = sizeof(thread_counts) / sizeof(thread_counts[0]);

    printf("Running multi-threaded shared variable with synchronization with %lld samples...\n", num_samples);
    printf("Running %d times per configuration...\n\n", num_runs);

    double baseline_time = 0.0;
    for (int t = 0; t < count_len; ++t) {
        int num_threads = thread_counts[t];
        double total_time = 0.0;
        double min_time = 1e9;
        double max_time = 0.0;
        double avg_pi = 0.0;

        for (int run = 0; run < num_runs; ++run) {
            pthread_t threads[num_threads];
            ThreadData data[num_threads];
            long long samples_per_thread = num_samples / num_threads;
            long long shared_inside = 0;
            pthread_mutex_t mutex;
            pthread_mutex_init(&mutex, NULL);

            for (int i = 0; i < num_threads; ++i) {
                data[i].samples = samples_per_thread;
                data[i].seed = 42u + (unsigned int)i + (unsigned int)run * 31u;
                data[i].shared_inside = &shared_inside;
                data[i].mutex = &mutex;
            }

            double start = get_seconds();
            for (int i = 0; i < num_threads; ++i) {
                pthread_create(&threads[i], NULL, worker, &data[i]);
            }
            for (int i = 0; i < num_threads; ++i) {
                pthread_join(threads[i], NULL);
            }
            double elapsed = get_seconds() - start;
            double estimated_pi = (double)shared_inside / (double)(samples_per_thread * num_threads) * 4.0;
            avg_pi += estimated_pi;
            total_time += elapsed;
            if (elapsed < min_time) min_time = elapsed;
            if (elapsed > max_time) max_time = elapsed;
            pthread_mutex_destroy(&mutex);
        }

        double avg_time = total_time / num_runs;
        avg_pi /= num_runs;
        double speedup = 0.0;
        double efficiency = 0.0;
        if (t == 0) {
            baseline_time = avg_time;
            speedup = 1.0;
            efficiency = 100.0;
        } else {
            speedup = baseline_time / avg_time;
            efficiency = speedup / num_threads * 100.0;
        }
        double accuracy = (1.0 - fabs(avg_pi - M_PI) / M_PI) * 100.0;
        printf("threads=%2d avg=%.6f min=%.6f max=%.6f speedup=%.4fx efficiency=%.2f%% accuracy=%.4f%%\n",
               num_threads, avg_time, min_time, max_time, speedup, efficiency, accuracy);
    }
    return 0;
}
